### PR TITLE
Add Helion fused cos/sin RoPE override

### PR DESCRIPTION
**Pull request by ethche**
Adds an opt-in override (`torchtitan/overrides/helion_rope.py`) that swaps `CosSinRoPE` for a version applying the cos/sin gather + rotate-half rotation in a single fused Helion kernel (forward and backward). Activated with `--override.imports torchtitan.overrides.helion_rope`; it never changes default behavior.

We have speedups for tok/s in a standard inference setting:

Qwen3-32B (TP-8) tok/s (higher is better)
Inference (vllm, eager + cudagraph) using https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/generate.py
| Prefix tokens | Input tokens | Output tokens | cos_sin tok/s | Helion tok/s | Speedup |
|---:|---:|---:|---:|---:|---:|
| 64 | 128 | 128 | 588.5 | 701.8 | **1.19x** |
| 64 | 1024 | 128 | 337.7 | 462.3 | **1.37x** |
| 64 | 128 | 512 | 661.4 | 755.3 | **1.14x** |

We see that the custom kernel improves eager fwd/bwd performance, but doesnt help performance under torch compile, due to inductor fusing the bwd kernel well + standard custom op overhead. We should think of this primarily as a inference/eager optimization.

Qwen3-32B (TP-8) end-to-end fwd_bwd timing (lower is better)

| B | S | Eager cos_sin ms | Eager helion ms | Eager speedup | Inductor cos_sin ms | Inductor helion ms | Inductor speedup |
|---:|---:|---:|---:|---:|---:|---:|---:|
| 8 | 1024 | 1157.1 | 1102.6 | 1.049x | 678.0 | 678.2 | 1.000x |
| 16 | 1024 | 1588.9 | 1490.6 | 1.066x | 1307.7 | 1309.9 | 0.998x |
| 32 | 1024 | 3129.1 | 2924.8 | 1.070x | 2559.7 | 2561.4 | 0.999x |
| 64 | 1024 | 6169.4 | 5770.2 | 1.069x | 5043.0 | 5054.0 | 0.998x |
| 8 | 4096 | 3308.2 | 3115.7 | 1.062x | 2753.1 | 2753.6 | 1.000x |
| 16 | 4096 | 6526.5 | 6118.3 | 1.067x | 5420.3 | 5426.4 | 0.999x |
| 32 | 4096 | 13005.8 | 12223.7 | 1.064x | 10824.7 | 10837.5 | 0.999x |
| 64 | 4096 | 25908.3 | 24331.7 | 1.065x | 21544.2 | 21558.9 | 0.999x |

Isolated kernel benchmarks (lower is better)

FWD RoPE
| B | S | eager fwd ms | inductor fwd ms | helion fwd ms | helion speedup vs inductor |
|---:|---:|---:|---:|---:|---:|
| 1 | 1024 | 0.1393 | 0.0663 | 0.0610 | 1.087x |
| 1 | 4096 | 0.1614 | 0.0676 | 0.0625 | 1.081x |
| 2 | 1024 | 0.1394 | 0.0654 | 0.0630 | 1.038x |
| 2 | 4096 | 0.3183 | 0.0725 | 0.0663 | 1.093x |
| 4 | 1024 | 0.1624 | 0.0675 | 0.0630 | 1.070x |
| 4 | 4096 | 0.5789 | 0.0852 | 0.0780 | 1.091x |
| 8 | 1024 | 0.3177 | 0.0696 | 0.0636 | 1.094x |
| 8 | 4096 | 1.0891 | 0.1222 | 0.1153 | 1.060x |
| 16 | 1024 | 0.5760 | 0.0801 | 0.0770 | 1.039x |
| 16 | 4096 | 2.1220 | 0.2020 | 0.1836 | 1.100x |
| 32 | 1024 | 1.0830 | 0.1143 | 0.1128 | 1.013x |
| 32 | 4096 | 4.1607 | 0.3462 | 0.3269 | 1.059x |
| 64 | 1024 | 2.1176 | 0.1823 | 0.1817 | 1.003x |
| 64 | 4096 | 8.2326 | 0.6416 | 0.6210 | 1.033x |
| 128 | 1024 | 4.1597 | 0.3376 | 0.3256 | 1.037x |
| 128 | 4096 | 16.3864 | 1.2524 | 1.2087 | 1.036x |

BWD RoPE
| B | S | Bwd eager ms | Bwd inductor ms | Bwd helion ms | Bwd speedup vs inductor |
|---:|---:|---:|---:|---:|---:|
| 1 | 1024 | 0.3177 | 0.1573 | 0.0730 | 2.157x |
| 1 | 4096 | 0.2915 | 0.1721 | 0.0698 | 2.466x |
| 2 | 1024 | 0.3263 | 0.1741 | 0.0763 | 2.282x |
| 2 | 4096 | 0.4542 | 0.1568 | 0.0714 | 2.197x |
| 4 | 1024 | 0.3225 | 0.1888 | 0.0693 | 2.726x |
| 4 | 4096 | 0.7800 | 0.1791 | 0.0859 | 2.084x |
| 8 | 1024 | 0.4382 | 0.1629 | 0.0687 | 2.371x |
| 8 | 4096 | 1.4483 | 0.1837 | 0.1233 | 1.490x |
| 16 | 1024 | 0.7838 | 0.1760 | 0.0855 | 2.059x |
| 16 | 4096 | 2.7736 | 0.2413 | 0.1901 | 1.269x |
| 32 | 1024 | 1.4500 | 0.1814 | 0.1217 | 1.490x |
| 32 | 4096 | 5.4222 | 0.4102 | 0.3306 | 1.241x |
| 64 | 1024 | 2.7465 | 0.2167 | 0.1927 | 1.124x |
| 64 | 4096 | 10.6831 | 0.7142 | 0.6191 | 1.154x |
| 128 | 1024 | 5.4158 | 0.4159 | 0.3323 | 1.251x |
| 128 | 4096 | 21.0995 | 1.3449 | 1.1790 | 1.141x |

Scope and safety:
- Adds an `exact` flag to the `@override` mechanism (`torchtitan/config/override.py`): a target matches subclasses by default; `exact=True` restricts to the concrete target type, which this override needs. Thus, it replaces exactly `CosSinRoPE.Config` and leaves subclasses with a different cache contract (e.g. Qwen3.5 `MRoPE`) untouched.
- Optional dependency: imports cleanly without `helion` and falls back to the PyTorch `CosSinRoPE` path when helion is absent or inputs are unsupported (non-CUDA, cross-device, non-4D or mismatched q/k, a cache that isn't a 2D `(max_seq, 2*head_dim)` table, non-integer or out-of-range position ids). Position ids are bounds-checked exactly as the PyTorch path.
- Compile-safe: the kernel is wrapped in a `torch.library.custom_op` with `register_fake`/`register_autograd`. Under compiled *training* the opaque custom op blocks Inductor's neighbor-fusion, so model-level compile can be a wash; this is documented rather than hidden behind a fallback.

### CODE REVIEW

File: torchtitan/overrides/helion_rope.py [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+
+else:
+
+    def _apply_helion_rope(
+        query: torch.Tensor,
+        key: torch.Tensor,
+        rope_cache: torch.Tensor,
+        positions: torch.Tensor | None,
+    ) -> tuple[torch.Tensor, torch.Tensor] | None:
+        warn_once(
```
Comment (tianyu-l) [inline torchtitan/overrides/helion_rope.py]: let's error out when user specify helion override but didn't install helion

File: tests/unit_tests/test_helion_rope.py
Comment (tianyu-l) [inline tests/unit_tests/test_helion_rope.py:1]: please add an integration test similar to https://github.com/pytorch/torchtitan/blob/main/tests/integration_tests/models.py#L118
where we apply
- Qwen3 debug model with FusedQKVLinear `qwen3_debugmodel_fused_qkv`
- FSDP2 TP2 CP2
- apply compile `--compile.enable`
- apply helion rope override

Comment (tianyu-l) [reply tests/unit_tests/test_helion_rope.py:1]: may need to install to CI environment
- if we should depend on helion stable package, then we could put it in https://github.com/pytorch/torchtitan/blob/main/.ci/docker/requirements-dev.txt
- if nightly, then similar to torchao https://github.com/pytorch/torchtitan/blob/main/.github/workflows/integration_test_8gpu_models.yaml#L78

**Review by tianyu-l — COMMENTED** [review]
For the inference speed up using vllm
- what is the set up? Are you using https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/generate.py to run torchtitan models using vllm inference engine?
- for "shape (prompt→out)", "decode | 192 → 512", could you briefly explain what it means? I'm guessing the shape change is due to cudagraph padding, but 256 sounds closer to pad to and maybe it's not selecting every power of 2 to pad to.


For the table on training perf, could you explain what's the unit?
> eager cos_sin | eager helion | eager Δ | compile cos_sin | compile helion | compile Δ
> -- | -- | -- | -- | -- | --

Are you saying helion is improving eager, but not compile, because helion op is opaque and prevents compiler fusion? If so,
- the numbers look confusing (maybe reversed?) if they mean throughput
- helion couldn't beat compiler fusion because the helion kernel we host is not for the tested shape?

**Comment by ethche** [comment]
> For the inference speed up using vllm
> 
> * what is the set up? Are you using https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/generate.py to run torchtitan models using vllm inference engine?
> * for "shape (prompt→out)", "decode | 192 → 512", could you briefly explain what it means? I'm guessing the shape change is due to cudagraph padding, but 256 sounds closer to pad to and maybe it's not selecting every power of 2 to pad to.
> 
> For the table on training perf, could you explain what's the unit?
> 
> > eager cos_sin
> > eager helion
> > eager Δ
> > compile cos_sin
> > compile helion
> > compile Δ
> 
> Are you saying helion is improving eager, but not compile, because helion op is opaque and prevents compiler fusion? If so,
> 
> * the numbers look confusing (maybe reversed?) if they mean throughput
> * helion couldn't beat compiler fusion because the helion kernel we host is not for the tested shape?

I apologize for the confusion:
* Yes, the current inference benchmark results are using the vllm inference engine specified in generate.py . Btw, I was using an older version of torchtitan, so I re-ran the benchmarks and updated with new numbers.
* "shape (prompt→out)", "decode | 192 → 512",  isn't related to cudagraph padding. It was referencing input tokens vs output tokens. So in this case it was 192 input tokens, 512 output tokens (which i called a "decode" shape, vs the prefill settings).
* For training perf, this was in milliseconds (lower was better). I also updated these numbers as well. We get an improvement over eager for fwd_bwd, but compared to inductor the performance is essentially identical. I suspect this is because (1) overhead from custom_op (2) inductor is able to fuse the bwd rope with neighbors well (3) our configs are not tuned completely, for the sake of having a simpler config setup.

File: torchtitan/overrides/helion_rope.py (hunk 1 above)
Comment (ethche) [reply torchtitan/overrides/helion_rope.py]: Great, it now raises an error.

File: tests/unit_tests/test_helion_rope.py
Comment (ethche) [reply tests/unit_tests/test_helion_rope.py:1]: Thank you. I put it in the yaml, but not requirements-dev.txt, in order to mirror the setup for torchao. Let me know though if I should but it in there anyways.

File: .github/workflows/integration_test_8gpu_models.yaml [hunk 2]
```diff
@@ -74,6 +74,13 @@ jobs:
         fi
 
         USE_CPP=0 python -m pip install --pre torchao --index-url ${{ matrix.index-url }}
+        # Helion is a PyPI-only pure-Python JIT package; install it after the
+        # nightly torch wheel and only on CUDA where the override integration runs.
+        if [[ "${{ matrix.gpu-arch-type }}" == "cuda" ]]; then
+          python -m pip install helion==1.1.0
+          python -c "import helion"
+          python -m pytest tests/unit_tests/test_helion_rope.py -v
```
Comment (tianyu-l) [inline .github/workflows/integration_test_8gpu_models.yaml]: Let's not run it here, as the CPU unit test would run everything under `unit_tests`.

Btw, is it runnable on CPU?

File: .github/workflows/integration_test_8gpu_models.yaml [hunk 3]
```diff
@@ -74,6 +74,13 @@ jobs:
         fi
 
         USE_CPP=0 python -m pip install --pre torchao --index-url ${{ matrix.index-url }}
+        # Helion is a PyPI-only pure-Python JIT package; install it after the
+        # nightly torch wheel and only on CUDA where the override integration runs.
+        if [[ "${{ matrix.gpu-arch-type }}" == "cuda" ]]; then
+          python -m pip install helion==1.1.0
```
Comment (tianyu-l) [inline .github/workflows/integration_test_8gpu_models.yaml]: If there's stable version, I'd prefer we put it in requirements-dev.

The reason torchao was put in yaml was because we need to depend on torchao nightly.

**Review by tianyu-l — COMMENTED** [review]
Thanks! Please rebase and address final nit comments.

File: .github/workflows/integration_test_8gpu_models.yaml (hunk 2 above)
Comment (ethche) [reply .github/workflows/integration_test_8gpu_models.yaml]: Sounds good, I removed it from here.

It falls back to pytorch cos_sin rope on CPU.

File: .github/workflows/integration_test_8gpu_models.yaml (hunk 3 above)
Comment (ethche) [reply .github/workflows/integration_test_8gpu_models.yaml]: Sounds good -- added to requirements-dev.

**Review by tianyu-l — APPROVED** [review]
LGTM, thank you!
